odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1471 @@
+                         convex=convex_conj_f.is_convex)
+
+        self._orig_convex_conj_f = convex_conj_f
+        self._scaling = float(scaling)
+
+    def _call(self, x):
+        """Applies the functional to the given point.
+
+        Returns
+        -------
+        `self(x)` : `float`
+            Evaluation of the functional.
+        """
+        return self._orig_convex_conj_f(x * (1 / self._scaling))
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluated
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return ((1 / self._scaling) * self._orig_convex_conj_f.gradient *
+                (1 / self._scaling))
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the ConvexConjugateArgScaling
+        functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator
+
+        Returns
+        -------
+        out : Operator
+            Domain and range equal to domain of functional
+        """
+        return proximal_arg_scaling(self._orig_convex_conj_f.proximal,
+                                    scaling=(1 / self._scaling))(sigma)
+
+    # TODO: Should it be added?
+    # Note: This would only be valide when f is proper convex and lower-
+    # semincontinuous.
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of the functional."""
+        raise NotImplementedError
+
+
+class ConvexConjugateLinearPerturb(Functional):
+    """ The ``Functional`` representing (F(.) + <y,.>)^*.
+
+    Calculate the convex conjugate functional perturbed function F(x) + <y,x>.
+
+    This is calculated according to the rule
+
+        (F(.) + <y,.>)^* (x) = F^*(x - y)
+
+    where ``y`` is the linear perturbation.
+
+    Parameters
+    ----------
+    convex_conj_f : `Functional`
+        Function corresponding to F^*.
+
+    y : Element in domain of F^*.
+
+    Notes
+    -----
+    For reference on the identity used, see [KP2015]_. Note that this is only
+    valide for functionals with a domain that is a Hilbert space.
+    """
+
+    def __init__(self, convex_conj_f, y):
+        """Initialize a ConvexConjugateLinearPerturb instance.
+
+        Parameters
+        ----------
+        convex_conj_f : `Functional`
+            Function corresponding to F^*.
+
+        y : Element in domain of F^*.
+        """
+        if y is not None and not isinstance(y, LinearSpaceVector):
+            raise TypeError('vector {!r} not None or a LinearSpaceVector'
+                            ' instance.'.format(y))
+
+        if y not in convex_conj_f.domain:
```

Comment:
I don't see a reason to allow `None` here since it's not a default value and makes no sense otherwise. My suggestion is to remove the `y is not None and` part from the first check (and the error message).